File: src/test/com/natpryce/snodge/internal/JsonPathTest.java
package com.natpryce.snodge.internal;

import com.google.gson.JsonElement;
import com.google.gson.JsonPrimitive;
import org.junit.Test;

import static com.natpryce.snodge.JsonBuilders.*;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.junit.Assert.assertEquals;

public class JsonPathTest
{
	@Test
	public void returnsSameHashCodeForSamePathNoMatterHowItIsConstructed() throws Exception
	{
		JsonPath path = JsonPath.of("a", 1, "b", 2, "c");

		assertEquals(path.hashCode(), JsonPath.root.extend("a").extend(1).extend("b").extend(2).extend("c").hashCode());
		assertEquals(path.hashCode(), JsonPath.root.extend("a", 1, "b", 2, "c").hashCode());
	}

	@Test
	public void pathsAreEqualNoMatterHowTheyAreConstructed() throws Exception
	{
		JsonPath path = JsonPath.of("a", 1, "b", 2, "c");

		assertEquals(path, JsonPath.root.extend("a").extend(1).extend("b").extend(2).extend("c"));
		assertEquals(path, JsonPath.root.extend("a", 1, "b", 2, "c"));
	}

	@Test
	public void canReplaceElementAtPath() throws Exception
	{
		JsonElement original = object(
				withField("a", list(
						object(
								withField("x", 10),
								withField("y", 20)),
						object(
								withField("x", 100),
								withField("y", 200)))),
				withField("b", "bubbles"));

		assertReplacement(original, JsonPath.of("a",1,"x"), new JsonPrimitive(-99), object(
				withField("a", list(
						object(
								withField("x", 10),
								withField("y", 20)),
						object(
								withField("x", -99),
								withField("y", 200)))),
				withField("b", "bubbles")));

		assertReplacement(original, JsonPath.root, new JsonPrimitive("just this"), new JsonPrimitive("just this"));
	}

	private void assertReplacement(JsonElement original, JsonPath path, JsonElement splice, JsonElement expected)
	{
		assertThat(path.replace(original, splice), equalTo(expected));
	}
}

File: src/test/com/natpryce/snodge/json/JsonPathTest.kt
package com.natpryce.snodge.json

import com.natpryce.hamkrest.assertion.assertThat
import com.natpryce.hamkrest.equalTo
import com.natpryce.jsonk.JsonElement
import com.natpryce.jsonk.JsonNumber
import com.natpryce.jsonk.JsonObject
import com.natpryce.jsonk.toJsonString
import com.natpryce.snodge.json.JsonPath.Companion.root
import org.junit.Test

class JsonPathTest {
    @Test
    fun `can query elements of path`() {
        val path = JsonPath("a", 1, "b", 2, "c")
        
        assertThat(path.size(), equalTo(5))
        assertThat(path[0], equalTo("a" as Any))
        assertThat(path[1], equalTo(1 as Any))
        assertThat(path[2], equalTo("b" as Any))
        assertThat(path[3], equalTo(2 as Any))
        assertThat(path[4], equalTo("c" as Any))
    }
    
    @Test
    fun `negative indices index from end of path`() {
        val path = JsonPath("a", 1, "b", 2, "c")
        
        assertThat(path[-5], equalTo("a" as Any))
        assertThat(path[-4], equalTo(1 as Any))
        assertThat(path[-3], equalTo("b" as Any))
        assertThat(path[-2], equalTo(2 as Any))
        assertThat(path[-1], equalTo("c" as Any))
    }
    
    @Test
    fun `returns same hash code for same path no matter how it is constructed`() {
        val path = JsonPath("a", 1, "b", 2, "c")
        
        assertThat(path.hashCode(), equalTo(root.extend("a").extend(1).extend("b").extend(2).extend("c").hashCode()))
        assertThat(path.hashCode(), equalTo(root.extend("a", 1, "b", 2, "c").hashCode()))
    }
    
    @Test
    fun `paths are equal no matter how they are constructed`() {
        val path = JsonPath("a", 1, "b", 2, "c")
        
        assertThat(path, equalTo(root.extend("a").extend(1).extend("b").extend(2).extend("c")))
        assertThat(path, equalTo(root.extend("a", 1, "b", 2, "c")))
    }
    
    @Test
    fun `can replace element at path`() {
        val original = obj(
            "a" to list(
                        obj(
                            "x" to 10,
                            "y" to 20),
                        obj(
                            "x" to 100,
                            "y" to 200)),
            "b" to "bubbles")
        
        assertReplacement(original, JsonPath("a", 1, "x"), JsonNumber(-99), obj(
            "a" to list(
                        obj(
                            "x" to 10,
                            "y" to 20),
                        obj(
                            "x" to -99,
                            "y" to 200)),
            "b" to "bubbles"))
        
        assertReplacement(original, JsonPath.root, JsonNumber("just this"), JsonNumber("just this"))
    }
    
    @Test
    fun `can replace element in singleton array at path`() {
        val original = list(1, 2, 3, list(1))
        
        assertReplacement(original, JsonPath(3, 0), JsonNumber(-99),
            list(1, 2, 3, list(-99)))
    }
    
    @Test
    fun `replacing an object field does not change order of elements when serialised`() {
        val original = obj(
            "a" to 1,
            "x" to 2,
            "b" to 3,
            "y" to 4)
        
        assertThat(original.toJsonString(), equalTo("{\"a\":1,\"x\":2,\"b\":3,\"y\":4}"))
        
        val replaced = JsonPath("x").replace(original, JsonNumber(99)) as JsonObject
        
        assertThat(replaced.toJsonString(), equalTo("{\"a\":1,\"x\":99,\"b\":3,\"y\":4}"))
    }
    
    private fun assertReplacement(original: JsonElement, path: JsonPath, splice: JsonElement, expected: JsonElement) {
        assertThat(path.replace(original, splice), equalTo(expected))
    }
}
